src/main/java/org/reactome/addlinks/referencecreators/TargetPathogenReferenceCreator.java

```diff
@@ -0,0 +1,219 @@
+				{
+					String[] mappedIdentifierParts = mapping.get(targetPathogenIdentifier).split("\\|");
+					for (String identifier : mappedIdentifierParts)
+					{
+						// If the string matches the Reactome pattern, it is a Reactome identifier.
+						if (REACTOME_IDENTIFIER_PATTERN.matcher(identifier).matches())
+						{
+							if (!reactomeToTargetPathogen.containsKey(identifier))
+							{
+								reactomeToTargetPathogen.put(identifier, new HashSet<>(Arrays.asList(targetPathogenIdentifier)));
+							}
+							else
+							{
+								Set<String> tmpSet = reactomeToTargetPathogen.get(identifier);
+								tmpSet.add(targetPathogenIdentifier);
+								reactomeToTargetPathogen.put(identifier, tmpSet);
+							}
+						}
+						else
+						{
+							if (!uniProtToTargetPathogen.containsKey(identifier))
+							{
+								uniProtToTargetPathogen.put(identifier, new HashSet<>(Arrays.asList(targetPathogenIdentifier)));
+							}
+							else
+							{
+								Set<String> tmpSet = uniProtToTargetPathogen.get(identifier);
+								tmpSet.add(targetPathogenIdentifier);
+								uniProtToTargetPathogen.put(identifier, tmpSet);
+							}
+						}
+					}
+				}
+			}
+		}
+		else
+		{
+			this.logger.info("Identifier sets appear to be already populated. Continuing with the process...");
+		}
+		this.logger.info("{} mappings in input (keyed by TargetPathogen identifier)", mapping.keySet().size());
+		this.logger.info("{} identifiers in reactome-to-TargetPathogen map", reactomeToTargetPathogen.size());
+		this.logger.info("{} identifiers in uniProt-to-TargetPathogen map", uniProtToTargetPathogen.size());
+		// Now we need to go through the sourceReferences and create references based on reactomeIdentifiers and uniProtIdentifiers.
+		// It might be necessary to have two instances of TargetPathogenReferenceCreator: one for creating cross-refs for Reactions and the other for creating cross-refs for Proteins.
+		// But... that would not be terribly efficient.
+
+		// So... which instance are we? We can determine from how the SourceRefDB and classReferring are set.
+		if (this.sourceRefDB.equals("UniProt") && this.classReferringToRefName.equals(ReactomeJavaConstants.ReferenceGeneProduct))
+		{
+			this.createReferencesForUniProtIdentifiers(sourceReferences);
+		}
+		else if (this.sourceRefDB.equals("Reactome") && this.classReferringToRefName.equals(ReactomeJavaConstants.Reaction))
+		{
+			this.createReferencesForReactomeIdentifiers(sourceReferences);
+		}
+		else
+		{
+			this.logger.error("sourceRefDB was: {} and classReferring was: {} - THIS IS NOT A VALID COMBINATION! Valid options are: UniProt/ReferenceGeneProduct and Reactome/Reaction.", this.sourceRefDB, this.classReferringToRefName);
+		}
+	}
+
+	/**
+	 * Create references for Reactome Identifiers.
+	 * @param sourceReferences
+	 * @throws Exception
+	 */
+	private void createReferencesForReactomeIdentifiers(List<GKInstance> sourceReferences) throws Exception
```

Comment:
This method and `createReferencesForUniProtIdentifiers` look almost identical.  Can the common logic be extracted out to remove the duplication?

Reply:
They differ in terms of which static-member cache they look things up in, as well as what value they use as as lookup key. They look similar, but there are subtle differences, and I'm not sure they are easy to refactor away, or that it's even worth the extra effort. Still, I'll give it another look...

Reply:
I was thinking something like this (starting from line 100 in the file):
```
	// So... which instance are we? We can determine from how the SourceRefDB and classReferring are set.
	if (this.sourceRefDB.equals("UniProt") && this.classReferringToRefName.equals(ReactomeJavaConstants.ReferenceGeneProduct))
	{
		this.createReferencesForIdentifiers(
			filterGKInstancesByClass(sourceReferences, ReactomeJavaConstants.ReferenceGeneProduct),
			this::getTargetPathogensFromUniProtSourceRef
		);
	}
	else if (this.sourceRefDB.equals("Reactome") && this.classReferringToRefName.equals(ReactomeJavaConstants.Reaction))
	{
		this.createReferencesForIdentifiers(
			filterGKInstancesByClass(sourceReferences, ReactomeJavaConstants.Reaction),
			this::getTargetPathogensFromReactomeStableIdentifierSourceRef
		);
	}
	else
	{
		this.logger.error("sourceRefDB was: {} and classReferring was: {} - THIS IS NOT A VALID COMBINATION! Valid options are: UniProt/ReferenceGeneProduct and Reactome/Reaction.", this.sourceRefDB, this.classReferringToRefName);
	}
}

private void createReferencesForIdentifiers(
	List<GKInstance> sourceReferences,
	Function<GKInstance, Set<String>> sourceRefToTargetPathogens
) throws Exception {
	int numRefsCreated = 0;
	int numRefsPreexisting = 0;
	int numTargetIdentifiers = 0;
	int[] counts = {numTargetIdentifiers, numRefsCreated, numRefsPreexisting};

	for (GKInstance sourceRef : sourceReferences) {
		for (String targetPathogenID : sourceRefToTargetPathogens.apply(sourceRef)) {
			counts = createReferencesForIdentifier(sourceRef, targetPathogenID, counts);
		}
	}

	this.logger.info("Inspected {} objects, {} identifiers for TargetPathogen.\n"
	+ " {} new references were created.\n"
	+ "{} references already existed (and were NOT re-created).",
	sourceReferences.size(), numTargetIdentifiers, numRefsCreated, numRefsPreexisting);
}

private List<GKInstance> filterGKInstancesByClass(List<GKInstance> gkInstances, String className) {
	return gkInstances.stream().filter(
		sr -> sr.getSchemClass().getName().equals(className)
	).collect(Collectors.toList());
}

private Set<String> getTargetPathogensFromUniProtSourceRef(GKInstance sourceRef) {
	Set<String> targetPathogens = new HashSet<>();
	try {
		String uniprotIdentifier = sourceRef.getAttributeValue(ReactomeJavaConstants.identifier).toString();
		targetPathogens = uniProtToTargetPathogen.computeIfAbsent(uniprotIdentifier, k -> new HashSet<>());
	} catch (Exception e) {
		this.logger.error("Unable to get UniProtIdentifier from {}", sourceRef.getExtendedDisplayName(), e);
	}

	return targetPathogens;
}

private Set<String> getTargetPathogensFromReactomeStableIdentifierSourceRef(GKInstance sourceRef) {
	Set<String> targetPathogens = new HashSet<>();

	try {
		GKInstance reactomeStableIdentifier = (GKInstance) sourceRef.getAttributeValue(ReactomeJavaConstants.stableIdentifier);
		String identifier = reactomeStableIdentifier.getAttributeValue(ReactomeJavaConstants.identifier).toString();
		targetPathogens = reactomeToTargetPathogen.computeIfAbsent(identifier, k -> new HashSet<>());
	} catch (Exception e) {
		this.logger.error("Unable to get Reactome Stable Identifier from {}", sourceRef.getExtendedDisplayName(), e);
	}

	return targetPathogens;
}
```

Reply:
Hmmm... interesting, but what problem does this large rewrite solve?

Reply:
Some of the logic that was duplicated between related methods gets consolidated.

Reply:
I still don't think the duplicate was _that_ big a deal. Still, I'll look into implementing something like this and then testing, if it's going to be a blocker. But this is a (relatively) large-ish rewrite, it will need to wait...

Reply:
Removing duplication is something I try to be pretty aggressive about.  It makes it easier if changes are needed so only one area has to be modified and logic doesn't get out of sync.

I try to stick as closely as I can to the DRY principle using simple methods so things are easier to unit test and reason about.